### PR TITLE
[12.0] [MIG] sale_operating_unit

### PR DESCRIPTION
**Pull request by bjeficent**
Migration of sale_operating_unit to version 12.0

### CODE REVIEW

File: sale_operating_unit/__manifest__.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
+# © 2015-17 Eficent Business and IT Consulting Services S.L.
+# - Jordi Ballester Alomar
```
Comment (elicoidal) [inline sale_operating_unit/__manifest__.py:2]: see https://github.com/OCA/operating-unit/pull/152 first as most of the remarks apply

File: sale_operating_unit/models/crm_team.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+    @api.multi
+    @api.constrains('operating_unit_id')
+    def _check_sales_order_operating_unit(self):
+        for rec in self:
+            orders = self.env['sale.order'].search(
+                [('team_id', '=', rec.id), ('operating_unit_id', '!=',
+                                            rec.operating_unit_id.id)])
+            if orders:
+                raise ValidationError(_('Sales orders already exist '
```
Comment (elicoidal) [inline sale_operating_unit/models/crm_team.py]: The message is not clear. Please rephrase

File: sale_operating_unit/models/sale.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+                self.operating_unit_id:
+            self.team_id = False
+
+    @api.multi
+    @api.constrains('team_id', 'operating_unit_id')
+    def _check_team_operating_unit(self):
+        for rec in self:
+            if (rec.team_id and
+                    rec.team_id.operating_unit_id != rec.operating_unit_id):
+                raise ValidationError(_('Configuration error. '
```
Comment (elicoidal) [inline sale_operating_unit/models/sale.py]: should not this be a title?

File: sale_operating_unit/models/sale.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+                                        'must match with that of the '
+                                        'quote/sales order'))
+
+    @api.multi
+    @api.constrains('operating_unit_id', 'company_id')
+    def _check_company_operating_unit(self):
+        for rec in self:
+            if (rec.company_id and rec.operating_unit_id and
+                    rec.company_id != rec.operating_unit_id.company_id):
+                raise ValidationError(_('Configuration error. The Company in '
```
Comment (elicoidal) [inline sale_operating_unit/models/sale.py]: should not this be a title?

File: sale_operating_unit/readme/CONTRIBUTORS.rst [hunk 5]
```diff
@@ -0,0 +1,7 @@
+
```
Comment (elicoidal) [inline sale_operating_unit/readme/CONTRIBUTORS.rst]: remove all starting lines in the fragments

File: sale_operating_unit/readme/DESCRIPTION.rst [hunk 6]
```diff
@@ -0,0 +1,5 @@
+
+This module was written to extend the Sales capabilities of Odoo.
```
Comment (elicoidal) [inline sale_operating_unit/readme/DESCRIPTION.rst]:
```suggestion
This module extends the Sales capabilities of Odoo
```

File: sale_operating_unit/readme/DESCRIPTION.rst [hunk 7]
```diff
@@ -0,0 +1,5 @@
...
+This module introduces the operating unit to the Sales Order.
```
Comment (elicoidal) [inline sale_operating_unit/readme/DESCRIPTION.rst]:
```suggestion
and  introduces the operating unit to the Sales Order object.
```

File: sale_operating_unit/readme/DESCRIPTION.rst [hunk 8]
```diff
@@ -0,0 +1,5 @@
...
+Security rules are defined to ensure that users can only display the
+Sales Orders in which they are allowed access to.
```
Comment (elicoidal) [inline sale_operating_unit/readme/DESCRIPTION.rst]:
```suggestion
Sales Orders they are allowed to access.
```

File: sale_operating_unit/views/sale_report_view.xml [hunk 9]
```diff
@@ -0,0 +1,22 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (elicoidal) [inline sale_operating_unit/views/sale_report_view.xml:1]: copyrights in XML files

**Review by elicoidal — CHANGES_REQUESTED** [review]

**Comment by elicoidal** [comment]
Travis is failing

**Review by CasVissers-360ERP — APPROVED** [review]
LGTM, functionally tested. Would maybe rename python filenames,

**Comment by CasVissers-360ERP** [comment]
The record rule "Sales Orders from allowed operating units" can cause errors for portal users since it is a global rule. Should there be an additional rule for portal users or should the rule only be applied to internal users? Or is the idea to give portal users rights to all operating units by default?

**Comment by AaronHForgeFlow** [comment]
@CasVissers Never thought about that, we need to think about tat. I guess the only solution now is give portal users access to all operating units.

File: sale_operating_unit/models/sale_order.py [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+    def _check_company_operating_unit(self):
+        for rec in self:
+            if (rec.company_id and rec.operating_unit_id and
+                    rec.company_id != rec.operating_unit_id.company_id):
+                raise ValidationError(_('Configuration error. The Company in '
+                                        'the Sales Order and in the Operating '
+                                        'Unit must be the same.'))
+
+    @api.multi
+    def _prepare_invoice(self):
```
Comment (CasVissers-360ERP) [inline sale_operating_unit/models/sale_order.py:61]: Should OU not be added to def _create_invoice on sale.advance.payment.inv as well?

Comment (AaronHForgeFlow) [reply sale_operating_unit/models/sale_order.py:61]: Yes, I think so. This improvement should be put also in previous versions

Comment (AaronHForgeFlow) [reply sale_operating_unit/models/sale_order.py:61]: > Should OU not be added to def _create_invoice on sale.advance.payment.inv as well?

Done. Thanks for letting me know.

**Comment by AaronHForgeFlow** [comment]
> The record rule "Sales Orders from allowed operating units" can cause errors for portal users since it is a global rule. Should there be an additional rule for portal users or should the rule only be applied to internal users? Or is the idea to give portal users rights to all operating units by default?

Could not reproduce. Portal user has read access to all OU.

**Review by AaronHForgeFlow — APPROVED** [review]
:+1: Functional

**Review by MiquelRForgeFlow — APPROVED** [review]
LGTM

**Comment by Dangar-Serpentcs** [comment]
I think here, need to change dependency **sale_management** instead of **sale** 
What do you think @nikul-serpentcs

**Comment by nikul-serpentcs** [comment]
> I think here, need to change dependency **sale_management** instead of **sale**
> What do you think @nikul-serpentcs

Yes brother without adding sale_management dependancy sale menu not visible, So should change **sale** to **sale_management**

CC @AaronHForgeFlow

**Comment by AaronHForgeFlow** [comment]
dependencies in __manifest__ are technical dependencies, but not usability dependencies. Thus, I would leave it as it is.

**Pull request merged**